File: collect_app/src/main/java/org/odk/collect/android/location/client/MapboxLocationCallback.java
package org.odk.collect.android.location.client;

import android.location.Location;

import androidx.annotation.NonNull;

import com.mapbox.android.core.location.LocationEngineCallback;
import com.mapbox.android.core.location.LocationEngineResult;

import org.odk.collect.android.geo.MapboxMapFragment;
import org.odk.collect.location.LocationUtils;

import java.lang.ref.WeakReference;

// https://docs.mapbox.com/android/core/guides/#requesting-location-updates
// Replace mock location accuracy with 0 as in LocationClient implementations since Mapbox uses its own location engine.
public class MapboxLocationCallback implements LocationEngineCallback<LocationEngineResult> {

    private final WeakReference<MapboxMapFragment> mapRef;
    private boolean retainMockAccuracy;

    public MapboxLocationCallback(MapboxMapFragment map) {
        mapRef = new WeakReference<>(map);
    }

    @Override
    public void onSuccess(LocationEngineResult result) {
        MapboxMapFragment map = mapRef.get();
        Location location = result.getLastLocation();
        if (map != null && location != null) {
            map.onLocationChanged(LocationUtils.sanitizeAccuracy(location, retainMockAccuracy));
        }
    }

    @Override
    public void onFailure(@NonNull Exception exception) { }

    public void setRetainMockAccuracy(boolean retainMockAccuracy) {
        this.retainMockAccuracy = retainMockAccuracy;
    }
}

File: mapbox/src/main/java/org/odk/collect/mapbox/MapboxLocationCallback.kt
package org.odk.collect.mapbox

import com.mapbox.android.core.location.LocationEngineCallback
import com.mapbox.android.core.location.LocationEngineResult
import org.odk.collect.location.LocationUtils.sanitizeAccuracy
import java.lang.Exception
import java.lang.ref.WeakReference

// https://docs.mapbox.com/android/core/guides/#requesting-location-updates
// Replace mock location accuracy with 0 as in LocationClient implementations since Mapbox uses its own location engine.
class MapboxLocationCallback(map: MapboxMapFragment) : LocationEngineCallback<LocationEngineResult> {
    private val mapRef: WeakReference<MapboxMapFragment> = WeakReference(map)
    private var retainMockAccuracy = false

    override fun onSuccess(result: LocationEngineResult) {
        val map = mapRef.get()
        val location = result.lastLocation
        if (map != null && location != null) {
            sanitizeAccuracy(location, retainMockAccuracy)?.let {
                map.onLocationChanged(it)
            }
        }
    }

    override fun onFailure(exception: Exception) = Unit

    fun setRetainMockAccuracy(retainMockAccuracy: Boolean) {
        this.retainMockAccuracy = retainMockAccuracy
    }
}
